langfuse/decorators/langfuse.py

```diff
@@ -0,0 +1,388 @@
+            - It updates the parameters of the most recently created observation on the observation stack. Care should be taken in nested observation contexts to ensure the updates are applied as intended.
+            - Parameters set to `None` will not overwrite existing values for those parameters. This behavior allows for selective updates without clearing previously set information.
+        """
+        stack = observation_stack_context.get()
+        observation = stack[-1] if stack else None
+
+        if not observation:
+            self.log.warn("No observation found in the current context")
+
+            return
+
+        observation_params_context.get()[observation.id].update(
+            input=input,
+            output=output,
+            name=name,
+            version=version,
+            metadata=metadata,
+            start_time=start_time,
+            end_time=end_time,
+            release=release,
+            tags=tags,
+            user_id=user_id,
+            session_id=session_id,
+            level=level,
+            status_message=status_message,
+        )
+
+    def flush(self):
+        """
+        Forces the immediate flush of all buffered observations to the Langfuse backend.
+
+        This method triggers the explicit sending of all accumulated trace and observation data that has not yet been sent to Langfuse servers.
+        It is typically used to ensure that data is promptly available for analysis, especially at the end of an execution context or before the application exits.
+
+        Usage:
+            - This method can be called at strategic points in the application where it's crucial to ensure that all telemetry data captured up to that point is made persistent and visible on the Langfuse platform.
+            - It's particularly useful in scenarios where the application might terminate abruptly or in batch processing tasks that require periodic flushing of trace data.
+
+        Returns:
+            None
+
+        Raises:
+            ValueError: If it fails to find a Langfuse client object in the current context, indicating potential misconfiguration or initialization issues.
+
+        Note:
+            - The flush operation may involve network I/O to send data to the Langfuse backend, which could impact performance if called too frequently in performance-sensitive contexts.
+            - In long-running applications, it's often sufficient to rely on the automatic flushing mechanism provided by the Langfuse client.
+            However, explicit calls to `flush` can be beneficial in certain edge cases or for debugging purposes.
+        """
+
+        langfuse = self._get_langfuse()
+        if langfuse:
+            langfuse.flush()
+        else:
+            self.log.warn("No langfuse object found in the current context")
+
+    def _get_langfuse(self) -> Langfuse:
```

Comment:
This here might be called concurrently / parallel. I guess we shoudl make sure this behaves like a singleton. We have a similar implementation in the openai integration. Feel free to check that out.